### PR TITLE
Updated Answer Key

### PR DESCRIPTION
**Pull request by Will-Hsu**
I added the output to the answer key!

### CODE REVIEW

File: content/english/java-tictactoe/activity-5.md [hunk 1]
```diff
@@ -3,5 +3,80 @@ title: "5. (Optional) AI Move"
 description: ""
 date: 2020-07-23T00:00:00Z
 weight: 6
-hidden: true
----
\ No newline at end of file
+hidden: false
```
Comment (Anvesha4) [inline content/english/java-tictactoe/activity-5.md:6]: Can we add a pre-req here with "java - functions", and maybe a difficulty level as Intermediate/Advanced?

File: content/english/java-tictactoe/answer-key.md [hunk 2]
```diff
@@ -230,7 +230,62 @@ public class TicTacToe_Nuevo {
 
 # activity-5: AI Computer Move
 ```java
-// TODO
+        public static int getComputerMoveAI(String[] curBoard){
```
Comment (Anvesha4) [inline content/english/java-tictactoe/answer-key.md:233]: It would be nice to have comments throughout the code for others to understand/debug/explain to the kids. Up to you if you want to do it as a separate PR

File: content/english/java-tictactoe/activity-5.md [hunk 3]
```diff
@@ -3,5 +3,80 @@ title: "5. (Optional) AI Move"
...
+- The algorithm makes optimal decision for the maximizer by examining all the possible future game states based on the current game board. 
+- If there is a winner or tie, the game state will be given a score. (Score: computer wins (`positive value`), player wins (`negative value`), tie (`0`)).
+- The two players are labeled as <b>maximizer</b> and <b>minimizer</b> respectedly. In this case, the computer is the maximizer, and the player is the minimizer. 
+- The maximizer tries to get the highest score possible, and the minimizer tries to minimize the score of its opponent. 
+
+<img src="../images/minimax.png" height="500"/> 
+
+During the minimizing round, we pick the game state with the smallest score. And, in the maximizing round, we pick the game state with the largest score.
+
+If you follow along through the maximizing/minimizing rounds on the picture above, you will note that the best move for the computer is to place `"X"` at position 8.
```
Comment (Anvesha4) [inline content/english/java-tictactoe/activity-5.md:32]: Do we need a picture with position labels?

File: content/english/java-tictactoe/activity-5.md [hunk 4]
```diff
@@ -3,5 +3,80 @@ title: "5. (Optional) AI Move"
...
+
+### Minimax Algorithm
+Minimax is an algorithm that is often used in two-player games to make optimal decisions for a player. 
+
+- The algorithm makes optimal decision for the maximizer by examining all the possible future game states based on the current game board. 
+- If there is a winner or tie, the game state will be given a score. (Score: computer wins (`positive value`), player wins (`negative value`), tie (`0`)).
+- The two players are labeled as <b>maximizer</b> and <b>minimizer</b> respectedly. In this case, the computer is the maximizer, and the player is the minimizer. 
+- The maximizer tries to get the highest score possible, and the minimizer tries to minimize the score of its opponent. 
+
+<img src="../images/minimax.png" height="500"/>
```
Comment (Anvesha4) [inline content/english/java-tictactoe/activity-5.md:28]: Love the picture, great job! 💯 

Consider a few modifications:
1. mark/separate the rows to clearly show they belong to different rounds so that students know they need to go top down (like 1st row: start, 2nd row and so on)
2. Maybe calculate scores for each of the moves & mark the ones that should be chosen that round? Not sure if that if that will make the picture crowded  - i'll leave it upto you to decide what to do about this

File: content/english/java-tictactoe/activity-5.md [hunk 5]
```diff
@@ -3,5 +3,80 @@ title: "5. (Optional) AI Move"
...
+
+We will do so by adding artificial intelligence into our program with the use the <b>Minimax</b> algorithm (a well-defined procedure that allows computers to solve problems).
+
+### Minimax Algorithm
+Minimax is an algorithm that is often used in two-player games to make optimal decisions for a player. 
+
+- The algorithm makes optimal decision for the maximizer by examining all the possible future game states based on the current game board. 
+- If there is a winner or tie, the game state will be given a score. (Score: computer wins (`positive value`), player wins (`negative value`), tie (`0`)).
+- The two players are labeled as <b>maximizer</b> and <b>minimizer</b> respectedly. In this case, the computer is the maximizer, and the player is the minimizer. 
+- The maximizer tries to get the highest score possible, and the minimizer tries to minimize the score of its opponent.
```
Comment (Anvesha4) [inline content/english/java-tictactoe/activity-5.md:26]: Would it be useful explaining how the score is calculated? I see we are using numbers in the picture below

File: content/english/java-tictactoe/activity-5.md [hunk 6]
```diff
@@ -3,5 +3,80 @@ title: "5. (Optional) AI Move"
...
+{{% notice hint %}}
+- Note that we call the method `minimax()` inside of the implementation of itself. We call a function that calls itself during its execution a <b>recursive function</b>.
+- Note you should pass `false` for the second argument since it'd be the minimizer's turn.
+- You should change that spot back to `"O"` after getting the score, so you maintain the original state of the game board during the next iteration.
+{{% /notice %}}
+
+2. Track the largest score and its corresponding position of the board at each iteration. Return the position with the largest score.
+{{% notice hint %}}
+- Have a variable called `bestScore` that stores the current best score and have an initial value of `Integer.MIN_VALUE` (minimum value of an integer).
+{{% /notice %}}
```
Comment (Anvesha4) [inline content/english/java-tictactoe/activity-5.md:49]: Consider using code snippets to explain some of the stuff here. The students still don't know the entire program and have to stitch these up anyway so that is still a challenge

File: content/english/java-tictactoe/activity-5.md [hunk 7]
```diff
@@ -3,5 +3,80 @@ title: "5. (Optional) AI Move"
...
+
+### Write the Method `minimax()`
+As we discussed above, the `minimax()` method has the header `int minimax(String[] curBoard, boolean isMaximizing)`.
+
+1. Call `getWinner()` on the board to check if there is a winner. If so, return the corresponding score (Score: computer wins (`1 * number of spots left + 1`), player wins (`-1 * number of spots left + 1`), tie (`0`)).
+
+{{% notice note %}}
+#### Why do we multiple the score by "number of spots left + 1"?
```
Comment (Anvesha4) [inline content/english/java-tictactoe/activity-5.md:57]: Not sure if you already talk about calculating score in previous activities.. if not, consider explaining this section with an example

**Comment by Anvesha4** [comment]
Great job, @Will-Hsu ! I have reviewed and added some comments inline. Please take a look and let me know if you have questions!

File: content/english/java-tictactoe/activity-5.md (hunk 3 above)
Comment (Will-Hsu) [reply content/english/java-tictactoe/activity-5.md:32]: I think we established the grid is labeled with 1-9 in the last 4 exercise, but I will add a note :)

File: content/english/java-tictactoe/activity-5.md (hunk 4 above)
Comment (Will-Hsu) [reply content/english/java-tictactoe/activity-5.md:28]: I change the arrows to make it clearer that each row is the different moves on can make! Thanks

File: content/english/java-tictactoe/activity-5.md (hunk 5 above)
Comment (Will-Hsu) [reply content/english/java-tictactoe/activity-5.md:26]: I moved the explanation to the top! Thanks

File: content/english/java-tictactoe/activity-5.md (hunk 1 above)
Comment (Will-Hsu) [reply content/english/java-tictactoe/activity-5.md:6]: Added 👍

File: content/english/java-tictactoe/activity-5.md (hunk 6 above)
Comment (Will-Hsu) [reply content/english/java-tictactoe/activity-5.md:85]: Added!

**Comment by Will-Hsu** [comment]
I made changes to activity 5 based on @Anvesha4 's comments! They are very helpful. (I still need to add comments to the answer key as you suggested tho~)

File: content/english/java-tictactoe/answer-key.md (hunk 2 above)
Comment (Will-Hsu) [reply content/english/java-tictactoe/answer-key.md:247]: added!

File: content/english/java-tictactoe/activity-5.md [hunk 8]
```diff
@@ -1,7 +1,111 @@
...
+
+### Minimax Algorithm
+Minimax is an algorithm that is used in two-player games to make optimal decisions for a player. 
+
+- The algorithm makes optimal decision for the maximizer by examining all the possible future game states based on the current game board. 
+- The two players are labeled as <b>maximizer</b> and <b>minimizer</b> respectedly. In this case, the computer is the maximizer, and the player is the minimizer. 
+- If there is a winner or tie, that game state will be given a score.
+- In our game specifically, we want the computer to win in fewest steps possible. Hence, we design the calculation of score for the game board as the following: 
+
+   (computer wins (`1 * number of spots left + 1`), player wins (`-1 * number of spots left + 1`), tie (`0`)).
```
Comment (Anvesha4) [inline content/english/java-tictactoe/activity-5.md:37]: In the first maximizing round, the score in the picture is +1*3 = 3, right?
According to the formula above, shouldn't it be 1*2+1 = 4, Since there are 2 spots left after the move?

Comment (Will-Hsu) [reply content/english/java-tictactoe/activity-5.md:37]: It's supposed to be 1 * (number of spots left + 1) => which make the score 1 *(2+1) = 1*3 = 3 since there are 2 spots left after the move. I will change the files to have parentheses!

File: content/english/java-tictactoe/activity-5.md [hunk 9]
```diff
@@ -1,7 +1,111 @@
...
+- activity-6: <a href="../../java-basics/activity-6" target="_blank">Object and Classes</a>
+- activity-7: <a href="../../java-basics/activity-7" target="_blank">Data Structures (Array & ArrayList)</a>
+{{% /notice %}}
+
+### Let's make the computer think
+Previously, we use a `Random` object to randomly generate a move for the computer. As a result, the competitiveness of the computer is fairly low.
+
+In this exercise, we want to increase the difficulty of the game by making optimal decisions for the computer.
+
+We will do so by adding artificial intelligence into our program with the use the <b>Minimax</b> algorithm (a well-defined procedure that allows computers to solve problems).
```
Comment (Anvesha4) [inline content/english/java-tictactoe/activity-5.md:27]: I feel there will be questions like why here - like why are we looking at so many moves, etc, and how etc. It will be nice to mention something like the following:

through our code, we will be looking into the future and calculating the score of all final states to make a decision at any given point. While it is difficult to do it mentally when we are playing the game ourselves, computers can do this very easily using something called _'recursion'_
Note that by giving scores to states, we are teaching our code what a win is, and how to reach that score..

File: content/english/java-tictactoe/activity-5.md [hunk 10]
```diff
@@ -1,7 +1,111 @@
...
+
+### Write the Method `getComputerMoveAI()`
+In activity-3, you wrote the method `int getComputerMove(String[] curBoard)` to randomly generate a spot for the computer. Let's write another method called `getComputerMoveAI(String[] curBoard)` that returns the optimal move for the computer. 
+
+We will write this method assuming we have the `int minimax(String[] curBoard, boolean isMaximizing)` method implemented, which you will write later.
+
+1. For each available spot on the game board, place `"O"` at that spot and get the score for that board by calling `minimax()`. 
+
+{{% notice hint %}}
+- Note that we call the method `minimax()` inside of the implementation of itself. We call a function that calls itself during its execution a <b>recursive function</b>.
```
Comment (Anvesha4) [inline content/english/java-tictactoe/activity-5.md:64]: Have we talked about recursion before? If not, we might have to add an advanced concept under function in basics

File: content/english/java-tictactoe/activity-5.md [hunk 11]
```diff
@@ -1,7 +1,111 @@
...
+   (computer wins (`1 * number of spots left + 1`), player wins (`-1 * number of spots left + 1`), tie (`0`)).
+- While the maximizer tries to get the highest score possible, the minimizer tries to minimize the score of its opponent.  
+
+Let's look at an example below:
+
+<img src="../images/minimax.png" height="500"/> 
+
+1. In the first row, we considered the 3 possible moves for `"O"`. 
+
+2. We note that if the computer puts its move at position 8, the game will halt as the computer wins. That state will hence be given a score of `1 * number of spots left + 1` = `8`.
```
[the remark this answers is not in the document]
Comment (Will-Hsu) [reply content/english/java-tictactoe/activity-5.md]: oops. It's supposed to say 3 ^

File: content/english/java-tictactoe/answer-key.md [hunk 12]
```diff
@@ -44,17 +47,21 @@ public class TicTacToe_Nuevo {
         while(true){
             int move;
 			try {
+                // get integer input
```
Comment (ozhang22) [inline content/english/java-tictactoe/answer-key.md:50]: It seems the comments don't line up exactly in Github because of spaces vs. tabs, can we make this consistent?

File: content/english/java-tictactoe/activity-5.md [hunk 13]
```diff
@@ -1,72 +1,120 @@
...
+1. In the first row, we considered the 3 possible moves for the computer `"O"`, who is the maximizer. 
 
-If you follow along through the maximizing/minimizing rounds on the picture above, you will note that the best move for the computer is to place `"X"` at position 8.
+2. We examine all the game states until all the moves end will a computer win, player win, or a tie. We then give them their corresponding score.
 
-### Write the Method `getComputerMoveAI()`
-In activity-3, you wrote the method `int getComputerMove(String[] curBoard)` to randomly generate a spot for the computer. Let's write another method called `getComputerMoveAI(String[] curBoard)` that returns the optimal move for the computer. 
+    For example, in the second board in row 1, the computer wins by placing `"O"` at position 8. That state will then be given a score of `1 * (number of available spots on the board + 1)` = `1 * (2+1)` = `3`.
+
+3. At game states that don't have a winner or a tie, we pick the smallest score during minimizing rounds (when `"X"` make a move), and the largest score during maximizing rounds (when `"X"` make a move).
```
Comment (ozhang22) [inline content/english/java-tictactoe/activity-5.md:55]:
```suggestion
3. At game states that don't have a winner or a tie, we pick the smallest score during minimizing rounds (when `"X"` make a move), and the largest score during maximizing rounds (when `"O"` make a move).
```
